bindings/rascal/models/kernels.py

```diff
@@ -52,19 +65,67 @@ def __init__(self, representation, name='Cosine', target_type='Structure',
             if not(zeta > 0 and zeta % 1 == 0):
                 raise ValueError(
                     "The given zeta has to be a positive integer.")
+        elif (name == 'GAP' and 'zeta' in kwargs):
+            # should be positive integer
+            zeta = kwargs['zeta']
+            if not(zeta > 0 and zeta % 1 == 0):
+                raise ValueError(
+                    "The given zeta has to be a positive integer.")
+        else:
+            raise RuntimeError("Kernel name must be one of: Cosine, GAP.")
         hypers = dict(name=name, target_type=target_type)
         hypers.update(**kwargs)
-        hypers_str = json.dumps(hypers)
+        self._rep = representation
         self._representation = representation._representation
+        self.name = name
+        self._kwargs = kwargs
+        self.kernel_type = kernel_type
+        self.target_type = target_type
+        if 'Sparse' in kernel_type:
+            self._kernel = SparseKernelcpp(hypers)
+        else:
+            self._kernel = Kernelcpp(hypers)
+
+    def get_init_params(self):
+        init_params = dict(representation=self._rep,
+                           name=self.name,
+                           kernel_type=self.kernel_type,
+                           target_type=self.target_type)
+        init_params.update(**self._kwargs)
+        return init_params
 
-        self._kernel = Kernelcpp(hypers_str)
+    def _set_data(self, data):
+        pass
 
-    def __call__(self, X, Y=None):
+    def _get_data(self):
+        return dict()
```

Comment:
yes there is no data to return here so the dict is empty but it still has to be a dict to be compatible with the to_dict() function in BaseIO. The output of to_dict() is a nested dictionary into which entries are either dictionaries or lists. It could also allow for `None` I guess but I don't see why it should. Is it clearer ?

Reply:
Yes - so to rephrase, it is because of the definition of the base class.

Reply:
you are right.